votor/src/consensus_pool/vote_pool.rs

```diff
@@ -1,129 +1,304 @@
+    ///
+    /// Per validator, we store a map of which block ids the validator has voted notar fallback on.
+    notar_fallback: BTreeMap<Pubkey, BTreeMap<Hash, VoteMessage>>,
 }
 
-impl SimpleVotePool {
-    pub(super) fn add_vote(
-        &mut self,
-        validator_vote_key: Pubkey,
-        validator_stake: Stake,
-        vote: VoteMessage,
-    ) -> Option<Stake> {
-        if !self.prev_voted_validators.insert(validator_vote_key) {
-            return None;
+impl Votes {
+    fn new(slot: Slot) -> Self {
+        Self {
+            slot,
+            skip: BTreeMap::default(),
+            skip_fallback: BTreeMap::default(),
+            finalize: BTreeMap::default(),
+            notar: BTreeMap::default(),
+            notar_fallback: BTreeMap::default(),
         }
-        self.votes.push(vote);
-        self.total_stake = self.total_stake.saturating_add(validator_stake);
-        Some(self.total_stake)
     }
 
-    pub(super) fn votes(&self) -> &[VoteMessage] {
-        &self.votes
-    }
-
-    pub(super) fn total_stake(&self) -> Stake {
-        self.total_stake
+    /// Adds votes.
+    ///
+    /// Checks for different types of slashable behavior and duplicate votes returning appropriate errors.
+    fn add_vote(&mut self, voter: Pubkey, vote: VoteMessage) -> Result<(), AddVoteError> {
+        assert_eq!(self.slot, vote.vote.slot());
+        match vote.vote {
+            Vote::Notarize(notar) => {
+                if self.skip.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.notar.entry(voter) {
+                    Entry::Occupied(e) => {
+                        // unwrap should be safe as we should only store notar type votes here
+                        if e.get().vote.block_id().unwrap() == &notar.block_id {
+                            Err(AddVoteError::Duplicate)
+                        } else {
+                            Err(AddVoteError::Slash)
+                        }
+                    }
+                    Entry::Vacant(e) => {
+                        e.insert(vote);
+                        Ok(())
+                    }
+                }
+            }
+            Vote::NotarizeFallback(nf) => {
+                if self.finalize.contains_key(&voter) {
+                    return Err(AddVoteError::Slash);
+                }
+                match self.notar_fallback.entry(voter) {
+                    Entry::Vacant(e) => {
+                        let mut map = BTreeMap::new();
+                        map.insert(nf.block_id, vote);
+                        e.insert(map);
+                        Ok(())
+                    }
+                    Entry::Occupied(mut e) => {
+                        let map = e.get_mut();
+                        let len = map.len();
+                        match map.entry(nf.block_id) {
+                            Entry::Vacant(map_e) => {
+                                if len == 3 {
```

Comment:
This 3 should be a constant defined somewhere (prefer it in common.rs, but I can accept it here)